src/dataflow/src/render/upsert.rs

```diff
@@ -150,36 +207,65 @@ where
                                             Ok(value) => {
                                                 if let Some(value) = value {
                                                     // prepend key to row
-                                                    row_packer.extend_by_row(&decoded_key);
-                                                    row_packer.extend_by_row(&value);
-                                                    Ok(Some(row_packer.finish_and_reuse()))
+                                                    let temp_storage = RowArena::new();
+                                                    // Unpack rows to apply predicates and projections.
+                                                    // This could be optimized with MapFilterProject.
+                                                    let mut datums =
+                                                        Vec::with_capacity(source_arity);
+                                                    datums.extend(decoded_key.iter());
+                                                    datums.extend(value.iter());
+
+                                                    match evaluate_predicates(
+                                                        &predicates,
+                                                        &datums[..],
+                                                        &temp_storage,
+                                                    ) {
```

Comment:
Interesting how arrangementless upsert means that we no longer need separate logic for predicate on only key columns.

Reply:
I think if we want we could identify predicates on keys and drop them entirely from the staging area, but we don't *need* to do that, which we might have before? Idk.